src/fetch/headers.js

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+'use strict';
+
+const { validateHeaderName, validateHeaderValue } = require('http');
+
+const { isPlainObject } = require('../common/utils');
+
+const INTERNALS = Symbol('Headers internals');
+
+const normalizeName = (name) => {
+  const nm = typeof name !== 'string' ? String(name) : name;
+
+  /* istanbul ignore next */
+  if (typeof validateHeaderName === 'function') {
+    // since node 14.3.0
+    validateHeaderName(nm);
+  } else {
+    // eslint-disable-next-line no-lonely-if
+    if (!/^[\^`\-\w!#$%&'*+.|~]+$/.test(nm)) {
+      const err = new TypeError(`Header name must be a valid HTTP token [${nm}]`);
+      Object.defineProperty(err, 'code', { value: 'ERR_INVALID_HTTP_TOKEN' });
+      throw err;
+    }
+  }
+
+  return nm.toLowerCase();
+};
+
+const normalizeValue = (value) => {
+  const val = typeof value !== 'string' ? String(value) : value;
+
+  /* istanbul ignore next */
+  if (typeof validateHeaderValue === 'function') {
+    // since node 14.3.0
+    validateHeaderValue('dummy', val);
+  } else {
+    // eslint-disable-next-line no-lonely-if
+    if (/[^\t\u0020-\u007E\u0080-\u00FF]/.test(val)) {
+      const err = new TypeError(`Invalid character in header content ["${val}"]`);
+      Object.defineProperty(err, 'code', { value: 'ERR_INVALID_CHAR' });
+      throw err;
+    }
+  }
+
+  return value;
+};
+
+/**
+ * Headers class
+ *
+ * @see https://fetch.spec.whatwg.org/#headers-class
+ */
+class Headers {
+  /**
+   * Constructs a new Headers instance
+   *
+   * @constructor
+   * @param {Object} [init={}]
+   */
+  constructor(init = {}) {
+    this[INTERNALS] = {
+      map: {},
```

Comment:
why not just use the `Map()` object? many functions entries/values will come for free

Reply:
Good point. However, while changing the implementation to use `Map` internally I remember why I did't in the first place: `Map` and its iterators are insertion ordered while the `Headers` iterators are alphabetically ordered. In the end the benefit of using `Map` versus `{}` is debatable. Never mind, here's the change: https://github.com/adobe/helix-fetch/pull/116/commits/2c13768b9fb9b7c095fd75c471d67f8d4fad5845